### ISSUE TITLE
[Question] Why gymnasium is pinned to 0.28.1 ?

### ISSUE THREAD

**Issue by duburcqa** [DLR-RM/stable-baselines3#1452]
### ❓ Question

Is enforcing `gymnasium==0.28.1` is strict requirement ? `ray[rllib]` has also pinned `gymnasium` but using a different version (`0.26.3`), so that installing both `stable-baseline3` and `ray[rllib]` at the same time is raising an exception. Would be be possible to relax this requirement, typically by allowing the range of versions `gynasium >= 0.26.3, <= 0.28.1` ?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
You can find your answer in https://github.com/DLR-RM/stable-baselines3/pull/1327 and https://github.com/DLR-RM/stable-baselines3/pull/780/files#r1116847655

TL;DR: The last year and a half has been a real pain in the neck for the SB3 devs, each new gym/gymnasium release came with breaking changes (more or less documented), so until gym is actually stable again, we have to pin to prevent any nasty surprises.
SB3 also requires the latest Gymnasium version for the type annotations (see discussions and bug found in the PR).

Note: it doesn't prevent you from installing ray by using `--no-deps`, I know it's not pretty but it works...

**Comment by duburcqa** [comment]
> each new gym/gymnasium release came with breaking changes

Yes I know that, that is why I suggest a range instead of a single version. It already pervents this kind of issues. If you tell me this is only for typing, this is not a concern for 99% of the users, so it is a bit sad to make the life harder for the all community because of this.

> Note: it doesn't prevent you from installing ray by using --no-deps, I know it's not pretty but it works...

Yes and no. It will not work in general because you actaully need to install deps most of the time.

> EDIT: I would rather push rllib to support latest Gymnasium version.

Ideally, but it is a company-owned project, so it is much harder to make it move forward.

**Comment by araffin** [comment]
> If you tell me this is only for typing, this is not a concern for 99% of the users, so it is a bit sad to make the life harder for the all community because of this. 

So, unfortunately, in that case, SB3 breaks with earlier version of Gymnasium (https://github.com/Farama-Foundation/Gymnasium/pull/337, https://github.com/DLR-RM/stable-baselines3/pull/1327/commits/ad48559b4e704f89896edb91bc3621e9236d267c): 
`TypeError: Too many parameters for <class 'gymnasium.core.Wrapper'>; actual 4, expected 2` 

> but it is a company-owned project, so it is much harder to make it move forward.

company-owned project should also mean more resources...
I don't recall all the breaking changes, but they should try simply upgrading to 0.28.1 if they are not using gym type annotations.

**Comment by duburcqa** [comment]
> So, unfortunately, in that case, SB3 breaks with earlier version of Gymnasium

If stable-baseline3 is failing at runtime for any other version of `gymnasium` than `0.28.1` then I agree there is no solution...

> company-owned project should also mean more resources...

Yes and no. They do, but only for what really matters for them and their clients, never for the open-source community.

**Comment by eccstartup** [comment]
```
$ pip install stable-baselines3
Collecting stable-baselines3
  Using cached stable_baselines3-1.8.0-py3-none-any.whl (174 kB)
Collecting gym==0.21 (from stable-baselines3)
  Using cached gym-0.21.0.tar.gz (1.5 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      error in gym setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by araffin** [comment]
> ```
> $ pip install stable-baselines3
> Collecting stable-baselines3
>   Using cached stable_baselines3-1.8.0-py3-none-any.whl (174 kB)
> Collecting gym==0.21 (from stable-baselines3)
>   Using cached gym-0.21.0.tar.gz (1.5 MB)
>   Preparing metadata (setup.py) ... error
>   error: subprocess-exited-with-error
> 
>   × python setup.py egg_info did not run successfully.
>   │ exit code: 1
>   ╰─> [1 lines of output]
>       error in gym setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed
> 
> × Encountered error while generating package metadata.
> ╰─> See above for output.
> 
> note: This is an issue with the package mentioned above, not pip.
> hint: See above for details.
> ```

please look at the issues in the repo, there are several duplicates of yours, and you are downloading SB3 1.8 not 2.x (currently in beta)

**Comment by araffin** [comment]
Closing as original question was answered and PR is on its way in RLLib repo: https://github.com/ray-project/ray/pull/35698
(SB3 master is also compatible with gymnasium 0.29)
